Named layouts (#4388)
sveltejs/kit: documentation/docs/02-layouts.md

<slot></slot>
```

### Named layouts

Some parts of your app might need something other than the default layout. For these cases you can create _named layouts_...

```svelte
/// file: src/routes/__layout-foo.svelte
<div class="foo">
	<slot></slot>
</div>
```

...and then use them by referencing the layout name (`foo`, in the example above) in the filename:

```svelte
/// file: src/routes/my-special-page@foo.svelte
<h1>I am inside __layout-foo</h1>
```

Named layouts are very powerful, but it can take a minute to get your head round them. Don't worry if this doesn't make sense all at once.

#### Scoping

Named layouts can be created at any depth, and will apply to any components in the same subtree. For example, `__layout-foo` will apply to `one` and `two`, but not `three` or `four`:

```
src/routes/
├ a/
│ ├ b/
│ │ ├ c/
│ │ │ ├ __layout-foo.svelte

> │ │ │ ├ one@foo.svelte

Reviewer: what if someone has a route like `folding@home.svelte` where they don't want the `@` to represent a named layout? should we add a mention here of some way to escape the `@` symbol or use route matchers?

Author: i've opened https://github.com/sveltejs/kit/issues/4403

```
src/routes/
├ a/
│ ├ b/
│ │ ├ c/
│ │ │ ├ __layout@foo.svelte
│ │ │ ├ one.svelte
│ │ │ ├ two.svelte
│ │ │ └ three.svelte
│ │ └ __layout.svelte
│ └ __layout.svelte
├ __layout.svelte
└ __layout-foo.svelte
```

Layouts that specify a parent will ordinarily ignore any default layouts, but you can force a named layout to inherit from the nearest default layout with the reserved `default` parent identifier. For example, a layout called `__layout-home@default.svelte` that contained a single `<slot/>` in the root directory would effectively enable any page, anywhere in the app, to inherit directly from the root layout, ignoring any intermediate layouts:

```
src/routes/
├ a/
│ ├ b/
│ │ ├ c/
│ │ │ ├ __layout.svelte                     # skipped
│ │ │ └ page-with-root-layout@home.svelte
│ │ └ __layout.svelte                       # skipped
│ └ __layout.svelte                         # skipped
├ __layout.svelte                           # inherited
└ __layout-home@default.svelte              # inherited

> ```

Author: I think it's important to include the description of what this enables (aka why you'd use it), including the explanation that it's cool and normal to have a layout that just contains a `<slot/>` for this purpose

> ```

Reviewer: Ok. I added that back to the code suggestion. The notion of "skipping" default layouts was just not quite how I thought of it as I think of it more as a lookup of the name. Not that either is wrong, but it's just a bit easier for me to conceptualize that way

> Layouts that specify a parent will ordinarily ignore any default layouts, but you can force a named layout to inherit from the nearest default layout with the reserved `default` parent identifier. For example, a layout called `__layout-home@default.svelte` that contained a single `<slot/>` in the root directory would effectively enable any page, anywhere in the app, to inherit directly from the root layout, ignoring any intermediate layouts:
> 
> ```
> src/routes/
> ├ a/
> │ ├ b/
> │ │ ├ c/
> │ │ │ ├ __layout.svelte                     # skipped
> │ │ │ └ page-with-root-layout@home.svelte
> │ │ └ __layout.svelte                       # skipped
> │ └ __layout.svelte                         # skipped
> ├ __layout.svelte                           # inherited
> └ __layout-home@default.svelte              # inherited
> ```

Author: FYI you can include code blocks in suggestions by incrementing the number of backticks, like so:

`````
````suggestion
### Some markdown

```js
// some code
```
````
`````

> ```

Author: (I edited your comment already)

> Layouts that specify a parent will ordinarily ignore any default layouts, but you can force a named layout to inherit from the nearest default layout with the reserved `default` parent identifier. For example, a layout called `__layout-home@default.svelte` that contained a single `<slot/>` in the root directory would effectively enable any page, anywhere in the app, to inherit directly from the root layout, ignoring any intermediate layouts:
> 
> ```
> src/routes/
> ├ a/
> │ ├ b/
> │ │ ├ c/
> │ │ │ ├ __layout.svelte                     # skipped
> │ │ │ └ page-with-root-layout@home.svelte
> │ │ └ __layout.svelte                       # skipped
> │ └ __layout.svelte                         # skipped
> ├ __layout.svelte                           # inherited
> └ __layout-home@default.svelte              # inherited
> ```

Author: tweaks:

````suggestion
Layouts that don't specify a parent layout will inherit from default (i.e. unnamed) layouts in parent directories. If you'd like to inherit from a default layout in the _current_ directory you can specify `@default`. For example, a layout called `__layout-home@default.svelte` that contained a lone `<slot/>` in the root directory would effectively enable any page, anywhere in the app, to inherit directly from the root layout — ignoring any intermediate layouts — by specifying `@home`:

```
src/routes/
├ a/
│ ├ b/
│ │ ├ c/
│ │ │ ├ __layout.svelte
│ │ │ └ page-with-root-layout@home.svelte
│ │ └ __layout.svelte
│ └ __layout.svelte
├ __layout.svelte                           # selected by `@default`
└ __layout-home@default.svelte              # selected by `@home`
```
````

> 

Reviewer 2: Not great englishing on my part, but this is probably a good place to take resets home (and make them searchable):

````suggestion
#### Layout resets

Taking into account scoping and inheritance rules, we can reset layouts to any depth we wish by carefully creating named layouts at the right position in the tree:

```
src/routes/
├ __layout.svelte
├ __layout-blank.svelte
└ a/
  ├ __layout@a.svelte
  ├ __layout-a.svelte
  └ b/
    ├ __layout.svelte
    └ c/
      ├ __layout.svelte
      ├ page-with-c's-layout.svelte
      ├ page-with-a's-layout@a.svelte
      └ page-with-empty-layout@blank.svelte
```

````

Author: I've been using this ropey-as-hell REPL app to generate the file tree layouts: https://svelte.dev/repl/5f66998f2c924a9eadb0954896682616?version=3.46.6. To make it consistent with the others it'd look more like this:

````suggestion
#### Layout resets

Taking into account scoping and inheritance rules, we can reset layouts to any depth we wish by carefully creating named layouts at the right position in the tree:

```
src/routes/
├ a/
│ ├ b/
│ │ ├ c/
│ │ │ ├ __layout.svelte
│ │ │ ├ page-with-c's-layout.svelte
│ │ │ ├ page-with-a's-layout@a.svelte
│ │ │ └ page-with-empty-layout@blank.svelte
│ │ └ __layout.svelte
│ ├ __layout@a.svelte
│ └ __layout-a.svelte
├ __layout.svelte
└ __layout-blank.svelte
```

````

Though I wonder if the `__layout-a`/`__layout@a` pair is a bit cryptic as things stand. Maybe something like this would be easier to follow? Though it uses a slightly different sorting to the other blocks, so they'd probably need to change. 

```
src/routes/
├ __layout.svelte        # (1)
├ __layout-root.svelte   # (2)
├ foo/
│ ├ __layout-foo.svelte  # (3)
│ ├ __layout@foo.svelte  # (4)
│ ├ bar/
│ │ ├ __layout.svelte    # (5)
│ │ ├ one.svelte         # 1 -> 3 -> 4 -> 5 -> one.svelte
│ │ ├ two@foo.svelte     # 1 -> 3 -> 4 -> two@foo.svelte
│ │ └ three@root.svelte  # 2 -> three@root.svelte
│ └ four.svelte          # 1 -> 3 -> 4 -> four.svelte
```

````

Reviewer 2: Yeah, and that's a use-case I would expect most people would never touch, so it was conflicting adding it.

Would inverting it to `__layout-foo@default.svelte` be easier or hard to grasp, I wonder? Or perhaps moving it a layer down like `/a/b/__layout-extends-a.svelte` (used `a` to be referenced at `c` since `/foo/bar` wouldn't be deep enough to get the point across)?

Author: I started writing "I think inverting it probably makes a ton of sense, since `@default` has already been discussed"...

````suggestion
#### Layout resets

Taking into account scoping and inheritance rules, we can reset layouts to any depth we wish by carefully creating named layouts at the right position in the tree:

```
src/routes/
├ __layout.svelte                # (1)
├ __layout-root.svelte           # (2)
├ foo/
│ ├ __layout.svelte              # (3)
│ ├ __layout-foo@default.svelte  # (4)
│ ├ bar/
│ │ ├ __layout.svelte            # (5)
│ │ ├ one.svelte                 # 1 -> 3 -> 5 -> one.svelte
│ │ ├ two@foo.svelte             # 1 -> 3 -> 4 -> two@foo.svelte
│ │ └ three@root.svelte          # 2 -> three@root.svelte
│ └ four.svelte                  # 1 -> 3 -> four.svelte
```

````

...but then we miss the fact that `four.svelte` inherits from a layout that inherits from the `foo` layout. Gah...

Reviewer 2: I'm not sure I follow—the example is supposed to be about resetting to layouts at arbitrary depths, and `__layout-foo@default` would indeed be an empty `<slot/>` anchor into `/foo/__layout`. Or at least that was my original intention with the "using `a`'s layout" example.

Reviewer 2: Although, yeah, `/foo/bar` isn't deep enough since `/foo/bar/__layout-alt.svelte` would have the same meaning.

It'd make more sense having `/foo/__layout-foo@default.svelte` and `/foo/bar/baz/page@foo.svelte` to hit the idea that `page` shares the same layout as `/foo/four.svelte`.

Author: just gonna go with the most recent suggestion for now

Author: oh hang on, missed your comments because github's UI sucks balls

└ __layout-foo.svelte
```

#### Inheritance chains

Layouts can themselves choose which other layouts they inherit from. Ordinarily, they will inherit from any default layouts 'above' them in the tree, but they can inherit named layouts using the same logic as pages. For example, `a/b/c/page.svelte` will inherit from `__layout-foo.svelte` and `a/b/c/__layout@foo.svelte` but _not_ `__layout.svelte` or `a/b/__layout.svelte`.

```
src/routes/
├ a/
│ ├ b/
│ │ ├ c/
│ │ │ ├ __layout@foo.svelte
│ │ │ └ page.svelte
│ │ └ __layout.svelte
│ ├ __layout.svelte
│ └ __layout-foo.svelte
```

Layouts will inherit from 'sibling' layouts as well as parents — you'll never need to do this, but you _could_:

```
src/routes/
├ __layout.svelte
├ __layout-a@.svelte
├ __layout-b@a.svelte
├ __layout-c@b.svelte
└ nested@c.svelte
```

Here, the inheritance chain would look like this (note that `__layout-a@` points at the unnamed default layout, because no name is specified):

```
__layout.svelte
  __layout-a@.svelte
    __layout-b@a.svelte
      __layout-c@b.svelte
        nested@c.svelte
```

#### Resetting layouts

> Sometimes you need a page to use the 'home' layout, regardless of its URL. You can do this by creating a top-level named layout that inherits from the default layout by using `@` without a name:

Reviewer: ```suggestion
You can create a named layout that does not inherit from others by using `@` without a name. E.g. you could do this to have a page use the 'home' layout, regardless of its URL:
```

Author: marking this resolved as the API has changed since this suggestion

```
src/routes/
├ a/
│ ├ b/
│ │ ├ c/
│ │ │ └ index@home.svelte
│ │ └ __layout.svelte
│ └ __layout.svelte
│ __layout.svelte

> └ __layout-home@.svelte

Reviewer: Could we just call it `__layout-home.svelte` instead?

Reviewer 2: So, reading your recent comments it seems we are still unclear about the semantics of inheritance and the `@` symbol.

What's unanimous so far is that `__layout.svelte` is a default layout, `__layout-foo.svelte` is a layout named `foo`, and `__layout[-foo]@bar.svelte` is a layout that extends the nearest layout named `bar`. 

The confusing parts are:

What do we mean by `__layout-foo@.svelte`? There are two options:

1) A layout `foo` that inherits from the nearest unnamed layout, aka. the default layout;
2) A layout `foo` that inherits from _no_ layout, aka. a layout reset.

Likewise, what do we mean by `__layout-foo.svelte`?

1) A layout that extends the nearest `foo`.
2) A layout that extends the nearest `foo` or, if that doesn't exist, the nearest default layout.

We must pick the same answer for both questions, as they are opposite sides of the same issue. As it stands, the first option is implemented.

We could switch to the second option, but I'm not sure it'd be more intuitive. For instance, with option 1, `__layout-foo.svelte` is equivalent to `__layout-bar@foo.svelte`. With option 2, you'd have too look up a parent layout `foo` differently depending on usage: `@foo` means `foo`, while `__layout-foo` means `foo` _or_ default layout.

I'd prefer option 1 and just work on rephrasing the docs. 

That said... I know @Rich-Harris dislikes reserved names, and we tried not to go there, but at this point I'm wondering if we should just _not_ support `@.svelte` at all and make `default` a reserved name for the default layout. It'd be clearer than picking a meaning for the empty shorthand.

Reviewer: > What do we mean by `__layout-foo@.svelte`?

If I read the docs correctly, then it's currently implemented as option 1. Layout resets are to be done using named layouts. I mostly agree with this, but question why `@` is needed at all in this case instead of just calling it `__layout-foo.svelte`

> Likewise, what do we mean by __layout-foo.svelte?

I'd go with option 1 here. And I'd probably have it throw an exception if `foo` does not exist.

Reviewer 2: > If I read the docs correctly, then it's currently implemented as option 1.

Yes, that's correct.

> I mostly agree with this, but question why `@` is needed at all in this case instead of just calling it `__layout-foo.svelte`

Because if we keep option 1, `__layout-foo.svelte` would be a reset instead of inheriting from the default layout, and in this case the intention would be to have `__layout-home@.svelte` inherit from `__layout.svelte` so we could reset to "the root layout" instead of an empty one.

Reviewer 2: I still believe the first option is easier to understand and, if we did consider reserving the `default` layout name, it'd be even easier. The whole ruleset would be:

- `__layout[-layout][@parent].svelte` defines a layout named `name = layout ?? 'default'` that extends the nearest layout named `parent ?? name`.
- `some-page[@parent].svelte` defines a page that extends the nearest layout named `parent ?? 'default'`.
- If `@parent` is explicitly specified it must be known, ie. `...@unknown.svelte` is an error.
- "Nearest layout named foo" means the first layout named `'foo'` found in the current folder or its parents.

In this case, `__layout-home@.svelte` wouldn't be a thing and we'd have to be explicit with `__layout-home@default.svelte`.

Reviewer: Ok. I think that makes sense and if we just clarify that in the docs it should probably be sufficient

Reviewer: I like being explicit with `default`

Author: closing as API was updated
